lib/core/src/Cardano/Wallet.hs

```diff
@@ -628,12 +635,17 @@ newWalletLayer tracer block0 feePolicy db nw tl = do
 
     _estimateTxFee
         :: DefineTx t
-        => WalletId
+        => BackendType
+        -> WalletId
         -> CoinSelectionOptions
         -> NonEmpty TxOut
         -> ExceptT ErrEstimateTxFee IO Fee
-    _estimateTxFee wid opts recipients = do
+    _estimateTxFee backend wid opts recipients = do
```

Comment:
I am afraid that this isn't an acceptable solution :s ! 
We carry the backend type as the type-level (that's what the `t` is for here, `t` is for `target` and represent the target backend type we are dealing with). We then use class-constraints to grant powers to a given function, making explicit what sort of backend-specific manipulation this function does. 

We can't have that at the value level though, as this will become rapidly unmanageable! In particular, code that is specific to a particular backend _must_ be located and contained into the module of that very backend. If we start putting in the shared piece (a.k.a core) things that are backend-specific this will spread and eventually become out of hand.

If I understand correctly, the issue here is to have some variations in the way null amounts are handled between either backend. Therefore, we need to provide a way for a particular implementation to tell whether a transaction / selection is valid or not. I'd suggest to extend the `TransactionLayer` with a new function: `validateSelection` or something similar which will yield either `()` or a `CreatedInvalidTransaction` (note that there's no need for a new branch `ErrEstimateTxFeeZeroAmountPresent`, which is by the way, maybe a bit too-specific ^^", since we already have an error that entirely captures that very issue). 

This function can then be implemented for both backends which have different rules about transactions. We know for instance that Jormungandr's transactions can't have more than 255 inputs or outputs  and that the bridge can't have output with a null amount. This function can then be called transparently from the wallet layer :), leaving the details out of the wallet core code!

Reply:
sound argument. Revisited all stuff!